src/uct/tcp/tcp_cm.c

```diff
@@ -21,69 +158,109 @@ unsigned uct_tcp_cm_conn_progress(uct_tcp_ep_t *ep)
         goto err;
     }
 
-    iface->outstanding--;
+    if (uct_tcp_cm_send_conn_req(ep) != UCS_OK) {
+        ucs_debug("tcp_ep %p: unable to send connection request to %s",
+                  ep, ucs_sockaddr_str(ep->peer_addr.addr, str_remote_addr,
+                                       UCS_SOCKADDR_STRING_LEN));
+        return 0;
+    }
 
-    uct_tcp_cm_change_conn_state(ep, UCT_TCP_EP_CONN_STATE_CONNECTED);
+    uct_tcp_cm_change_conn_state(ep, UCT_TCP_EP_CONN_STATE_WAITING_ACK);
+    uct_tcp_ep_mod_events(ep, EPOLLIN, EPOLLOUT);
 
     ucs_assertv((ep->tx.length == 0) && (ep->tx.offset == 0) &&
                 (ep->tx.buf == NULL), "ep=%p", ep);
-
-    return uct_tcp_ep_progress(ep, UCT_TCP_EP_CTX_TYPE_TX);
+    return 0;
 
 err:
-    iface->outstanding--;
     uct_tcp_ep_set_failed(ep);
     return 0;
 }
 
-void uct_tcp_cm_change_conn_state(uct_tcp_ep_t *ep, uct_tcp_ep_conn_state_t new_conn_state)
+unsigned uct_tcp_cm_conn_ack_rx_progress(uct_tcp_ep_t *ep)
 {
-    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface, uct_tcp_iface_t);
-    char str_local_addr[UCS_SOCKADDR_STRING_LEN], str_remote_addr[UCS_SOCKADDR_STRING_LEN];
-    uct_tcp_ep_conn_state_t old_conn_state;
+    ucs_status_t status;
 
-    old_conn_state = ep->conn_state;
-    ep->conn_state = new_conn_state;
+    status = uct_tcp_cn_recv_conn_ack(ep);
```

Comment:
return value from all progress functions should be > 0 if something handled. For example,
```
unsigned ret = 0;
...
ret += (status == UCS_OK);
...
return ret + uct_tcp_ep_progress(ep, UCT_TCP_EP_CTX_TYPE_TX);
```

Reply:
Good catch! Thank you!
This fixes the `test_ucp_wakeup.tx_wait` test.

Reply:
Unfortunately, it fixes the test on x86_64 only. Looks like this is timing related issue. Returned back "skip" for that test.